### ISSUE TITLE
Ram fills up

### ISSUE THREAD

**Issue by Rozin14** [emoose/VideoRenderer#1]
Ram fills up when playing 1080p videos on 2560*1600 screen. At whatever quality setting. Device: Laptop with 32GB and 3070 gpu

**Comment by emoose** [comment]
Confirmed on my side too, do you know what codec the video you tried with uses? it seemed to only happen with x265 for me, x264 1080p worked fine without memory issue.
(too bad browsers don't support x265, would have been neat to check with them)

MPC-HC also crashes if I try closing after the memory issue, and mentions nv_disp having an error, might be the issue is on NV's side, possible they never tried with x265 since browsers don't use it.

**Comment by emoose** [comment]
> You can disable P010 output in LAV Video Decoder. You also need to change hardware decoder from Native to Copyback.

Aha looks like that did the trick, video that ballooned to 10GB+ RAM now stays at 300MB or so, using NV12 instead of P010, thanks for the info!

Guess this is could be worth reporting to NV, not sure how amused they'll be about a bug report for this though, but hopefully they'll at least try looking into it.

>Some suggested code:

Ah good idea, will look into adding it soon.

---

For anyone else having this, you can access LAV Video decoder settings in MPC-HC's `Options > Internal Filters` section, just uncheck P010 & P016, and change Hardware decoder to `DXVA2 (copy-back)`

Didn't have much luck with MPC-BE's internal filter, but you can install LAV Filters from https://github.com/Nevcairiel/LAVFilters/releases and follow the same steps in https://www.youtube.com/watch?v=GaQvHLRVxgs to add them in, and use the `LAV Video Configuration` program that was installed to set the formats as above.

**Comment by clsid2** [comment]
D3D11 decoder uses Native by default. But you can let it do Copyback by selecting a hardware device below it.

Test if it works in your browser with HDR videos on Youtube. Those files use 10-bit VP9 or AV1 video.

**Comment by emoose** [comment]
Hm not noticing much with HDR videos, GPU usage stays flat with them, couldn't notice any difference toggling it neither, maybe they disable those in browsers somehow.

>Additionally, you can also use this for vendor specific code segments:

Added some checks for that earlier in https://github.com/emoose/VideoRenderer/commit/41c757e934f3f16c701adb1fa2b5bfc46f0b75f7 :)

What I'd really like to find is some way to tell if SuperRes is actually active, since atm the code there would display `Intel-VPE` / `Nvidia-VSR` in stats overlay even if it's not actually being processed, or user turned it off in NVCP etc...

Doesn't look like `VideoProcessorSetOutputExtension` gives any useful result code, might try messing with `GetOutputExtension` to see what that returns, otherwise maybe will change that text to "Nvidia-VSR-possible" or something like that.

**Comment by Rozin14** [comment]
YES. confirmed issues only with h265 videos. Fix of clsid2 does solve it

**Comment by emoose** [comment]
@clsid2 added your suggestions in https://github.com/emoose/VideoRenderer/commit/4e2e1605b903f1886d1c77922d1f4b66d3368be2 - seemed VSR was still working on some interlaced clips I tested though so removed the interlaced check.

Made it only check for P010/P016 too since we know those are likely faulty, if it's found that other formats are also having issues we can make it only allow NV12 though.

**Comment by clsid2** [comment]
Can you guys test with YUY2 and RBG32 input?
Simply uncheck all but one output format in LAV Video Decoder to force a specific colorspace.

**Comment by emoose** [comment]
Oh good idea, with YUY2 it still requests VSR but doesn't look like it's active, low power draw & no difference toggling in NVCP.

RGB32 didn't seem to request it, doesn't look like MPCVR is using VP for that since it mentions using shaders, wonder why...

**Comment by emoose** [comment]
>Your release notes mentioned that dithering should be disabled. Is that really true? Any idea why this breaks VSR? You might have used 10-bit video when testing this.

I noticed it was stopping VSR from working early on but not sure what video I was trying that with, just gave it another try and seemed to work fine with each video I tried now, guess that part can be removed to simplify the install.

Also in early tests 10/16-bit formats didn't seem to be working, the release page did mention changing it to 8-bit initially, but after trying some other videos it seems to be fine with any of the texture formats MPC-VR allows choosing, so removed that part from it.

**Comment by clsid2** [comment]
Next step to test is input size and aspect ratio limitations. What I have read so far VSR works for video with size up to and including 1080p. But maybe real max is 1440p on 4k screen. I also saw somewhere that it might now work for vertical video (height>width).

A good way to test with any size you want is to use ffdshow. Install ffdshow. You only need its raw video processing filter. Add that as a preferred external filter in MPC-HC. This filter has a resizing option where you can enter any desired size. So easy to test for example something slightly larger than 1080p, or to create a vertical video, with any input file you want.

`m_srcRect` can be used for input size in MPCVR.

**Comment by emoose** [comment]
Not having much luck with ffdshow using rev4533 it seems, added the "ffdshow raw video filter" as preferred external filter and enabled resizing in the ffdshow config, tested at 640x480 but video looks the same, and MPC-VR stats still shows the original video resolution there...

Tried testing a 4K video with desktop resolution at 5120x2880 via NV DSR, with superres disabled hwinfo showed it at ~80w, enabling it bumped it up to ~96w, toggling it in NVCP seemed to increase/drop the usage as I clicked it so guessing something is being activated there, but really couldn't see any difference in the output (image diff tools show them as identical too...)

**Comment by clsid2** [comment]
ffdshow should show up here when active:
MPC-HC menu > Play > Filters
It won't get used with DXVA2 Native. It also has a tray icon when it is active.

You can use Ctrl+E to quickly reload a video in MPC-HC. That may help if dynamic size changes do not work.

I doubt it is doing something useful with such minor power increase. Certainly for 4k input.

**Comment by emoose** [comment]
Maybe it's `NV12`, noticed it wasn't listed in the filters media types for ffdshow, if I change LAV to use `YUY2` that does let it take effect & shows in the Play > Filters menu, but with NV12 it doesn't show there any more, and we know that `YUY2` doesn't actually work with VSR :/

Tried adding guid for it to the mediatypes but that didn't seem to help neither, I noticed if I start playback with YUY2 and then change filters to use NV12 it does seem to keep the resize & MPC-VR says it's NV12 too, but kinda doubtful things are all being applied properly with that.

**Comment by clsid2** [comment]
Try with YV12 input and let ffdshow output NV12

**Comment by emoose** [comment]
Aha nice looks like it's working now, ty! Now to see what sizes it works with.

**Comment by clsid2** [comment]
Barely any difference. What is the extra power draw?

**Comment by kolunio82** [comment]
My Mpc hc player disabled HDR in windows cp all the time when I start playing video. I cant have HDR enabled in windows cp?

**Comment by emoose** [comment]
>Barely any difference. What is the extra power draw?

About 10w, went from ~87w to ~97w when toggling it.

Might be even worse with the older cards though, saw some people saying it can reach 200w+ for them but mines never gone near that high.

**Comment by clsid2** [comment]
It is the renderer that disables HDR, not the player. The option in the renderer controls it.

**Comment by kolunio82** [comment]
> It is the renderer that disables HDR, not the player. The option in the renderer controls it.

Ok but it is necessary? In MadVR i can enable hdr in windows cp and play any video.

**Comment by clsid2** [comment]
No, you adjust renderer options to not disable it. But it can potentially give you incorrect colors.

**Comment by kolunio82** [comment]
So in lav:
nv12 and p010 - unchecked others
ditherng - ordered or random?
in mpc:
chroma - bilinear
upscaling - catmul
downscaling - hamming
Correct?

**Comment by clsid2** [comment]
NV12 checked. If you want to process 10-bit video you need to **un**check P010 and the other 10/16-bit formats in LAV.

Leave the settings under "Shader video processor" as default. Those scaling algorithms are NOT used because scaling is done by DXVA2/D3D11 video processor!

**Comment by Pikkuhippi** [comment]
@kolunio82 
If you mean that your Windows HDR turns off when you playback videos with the renderer, change the setting:
"Auto display HDR On/Off -> not used"

**Comment by emoose** [comment]
Seems like 531.26 hotfix might not be applying VSR to P010 at all anymore, so ram issue doesn't happen, can anyone else who had the ram issue before confirm? (you'd need to use the older RTX MPC-VR that didn't check for P010 to test: https://github.com/emoose/VideoRenderer/releases/tag/rtx-1.0)
For me playing P010 gave no increased power/gpu usage, and no memory issue.

So for P010 you'd need to do the same as what we suggest here, using LAV to convert P010 to NV12 instead. Still hopeful they'll add proper support for P010 in future (seeing as VSR did seem to apply fine to it before besides the memory leak...)

**Comment by kolunio82** [comment]
> Seems like 531.26 hotfix might not be applying VSR to P010 at all anymore, so ram issue doesn't happen, can anyone else who had the ram issue before confirm? (you'd need to use the older RTX MPC-VR that didn't check for P010 to test: https://github.com/emoose/VideoRenderer/releases/tag/rtx-1.0) For me playing P010 gave no increased power/gpu usage, and no memory issue, but NV12 still worked fine.
> 
> So for P010 you'd still need to do the same as what we suggest here, using LAV to convert P010 to NV12 instead. Still hopeful they'll add proper support for P010 in future (seeing as VSR did seem to apply fine to it before besides the memory leak...)
> 
> Hotfix links can be found here: https://www.reddit.com/r/nvidia/comments/11l92kg/geforce_hotfix_driver_version_53126/

So how can I configure lav exactly?

**Comment by emoose** [comment]
>So how can I configure lav exactly?

It's mentioned in bottom part of https://github.com/emoose/VideoRenderer/issues/1#issuecomment-1450964365

**Comment by kolunio82** [comment]
When i enable hdr in wibdows, not mpc hc, settinga vsr don't work. I need to have hdr disabled in windows all the time? :(

**Comment by emoose** [comment]
@kolunio82 try the rtx-1.2 build at https://github.com/emoose/VideoRenderer/issues/5#issuecomment-1452590217, seemed to help others with the same issue.

**Comment by clsid2** [comment]
ProsRes is usually 10-bit which means you have to use the P010 workaround.

**Comment by kolunio82** [comment]
> ProsRes is usually 10-bit which means you have to use the P010 workaround.

So what shoud I do exactly? Im sorry for noob question. Im not advanced user.

**Comment by marcmy** [comment]
Hi, are these fixes still required for P010 and HDR still required? I just got the latest nvidia driver update (545.84) which enabled support for 20 series. Just checking if there were any changes or hotfixes made since this was last released. Thanks.

**Comment by hani3887gg** [comment]
can i use vsr on a 4k video?
